### ISSUE TITLE
mailctl in a container help

### ISSUE THREAD

**Issue by dngray** [pdobsan/mailctl#37]
I'm trying to get this working in a simple alpine linux container:

```Dockerfile
FROM docker.io/alpine:latest
ARG BUILD_DATE
ARG UID
LABEL name=mailctl
LABEL version="$IMAGE_VERSION"
LABEL created="$BUILD_DATE"

EXPOSE 8080

RUN adduser -D -s /bin/ash -u ${UID} mailctl

RUN echo "@testing https://dl-cdn.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories

RUN apk add --no-cache \
  gopass \
  mailctl@testing

USER mailctl

ENTRYPOINT ["tail", "-f", "/dev/null"]
```

Then the run commands:
```make
podman build -f mailctl/Containerfile \
	--build-arg BUILD_DATE=$(date -u) \
	--build-arg UID=1000 \
	-t mailctl:latest

podman network create --subnet 10.89.0.0/16 \
	--ipv6 --subnet fcaa:1::/48 mailctl_frontend

podman run -it --replace --userns=keep-id \
	-v ~/.config/mailctl:/home/mailctl/.config/mailctl:z \
	-p 8081:8080 \
	--net=mailctl_frontend --ip 10.89.0.2 \
	--name mailctl \
	localhost/mailctl
```

My services file is pretty much the same as the one in the [sample](https://github.com/pdobsan/mailctl/blob/main/configs/services-template.yaml#L20) for the microsoft account.

Inside the container I am able to run netcat ie `nc -l 10.89.0.2 -p 8080` and connect to it with `telnet localhost 8081`, but I'm unable to use mailctl on that port/configuration. I would have thought `redirect_uri: http://10.89.0.2:8080` would have worked, then I could have accessed it on `https://localhost:8081` in my browser on the host. (not inside the container).

Even when I try inside the container with `wget -q -O - http://10.89.0.2:8080/start` or `wget -q -O - http://127.0.0.1:8080/start` I get connection refused.

**Comment by dngray** [comment]
I think the issue might be that [`localhost` is hardcoded](https://github.com/pdobsan/mailctl/blob/main/lib/MailCtl/Authorization.hs#L338).

```haskell
localWebServer :: MVar AuthResult -> Environment -> String -> EmailAddress -> Bool -> IO ()
localWebServer mvar env serv email_ noHint = do
  generateAuthPage env serv email_ noHint
    >>= \case
      Left errmsg -> error $ "localWebServer:\n" ++ errmsg
      Right authP -> do
        let startAuth :: TW.ResponderM a
            startAuth = TW.send $ TW.html $ fromStrict authP
            localWebServerPort = getPortFromURIStr $ serviceFieldLookup (services env) serv redirect_uri
            localhostWebServer = Warp.setPort localWebServerPort $ Warp.setHost "localhost" Warp.defaultSettings
```

and this should perhaps be a configurable option maybe.

**Comment by pdobsan** [comment]
On Sat, Feb 10, 2024 at 11:07:26PM -0800, Daniel Nathan Gray wrote:
> I'm trying to get this working in a simple alpine linux container:

Interesting, that kind of use did not occur to me; but why not?

I can (in fact, should have from the beginning) make the webservers'
host parameter a real one instead of hard-wiring "localhost". I'll see
to it in "oama" the upcoming replacement for mailctl, see discussion #26.

Meanwhile, you have two options, most likely none of them perfect for
your use case.

> Works inside the container, but not from the outside.

You mean by "works inside" that you run a browser inside the container
while using its GUI on the host outside? I guess it is possible with
some kind of X11 network connection.

Alternatively, you perform the authorization once outside then you can
build your refresh token obtained into the container, it does not change
later.

**Comment by dngray** [comment]
> You mean by "works inside" that you run a browser inside the container while using its GUI on the host outside? I guess it is possible with some kind of X11 network connection.

No, I just used wget ie `wget -q -O - http://localhost:8080/start`. I don't want to have X running inside the container, or a fully fledged browser, that will pull a huge amount of dependency, size and surface area.

> Alternatively, you perform the authorization once outside then you can
build your refresh token obtained into the container, it does not change
later.

That's probably the most viable option. Alternatively I did come across [`pizauth`](https://github.com/ltratt/pizauth), which seems to fit a similar job to what mailctl does. I think it might have the same issue though as the author suggests using a [local ssh tunnel](https://github.com/ltratt/pizauth?tab=readme-ov-file#pizauth-on-a-remote-machine), although this [indicates there is a config option](https://github.com/ltratt/pizauth/blob/master/src/config.rs#L190).

**Comment by pdobsan** [comment]
> I can (in fact, should have from the beginning) make the webservers'
host parameter a real one instead of hard-wiring "localhost". 

No, I was too fast to jump to this conclusion!  Due to the security considerations there is no good reason to listen anything else but on `localhost` during *authorization*. If one, for whatever reasons, needs to perform *authorization*  on a remote host the solution is to use an ssh tunnel (`ssh -L ...`).

*Authorization* usually is performed only once since the _refresh token_ rarely needs to be changed if ever.

**Comment by dngray** [comment]
That does kind of make sense.

Anyway seems like I can't use this or `pizauth` as it doesn't support IMAP/SMTP anyway and I'll have to use [DavMail](https://davmail.sourceforge.net/)

https://man.sr.ht/~rjarry/aerc/providers/microsoft.md#office365-with-imap-disabled
